### PR TITLE
Use GGX for all microfacet shading models part I

### PR DESCRIPTION
**Pull request by est77**
Issue #2550

### CODE REVIEW

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 1]
```diff
@@ -589,13 +585,8 @@ namespace
             };
 
             shading_system.register_closure(name(), id(), params, &prepare_closure, nullptr);
-
             g_closure_convert_funs[id()] = &convert_closure;
-
             g_closure_get_modes_funs[id()] = &modes;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Do we still need `g_closure_get_modes_funs`?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 2]
```diff
@@ -283,100 +256,31 @@ namespace
...
-
-              default:
-                assert(!"Unexpected MDF type.");
-                wi = Vector3f(0.0f);
-                probability = 0.0f;
-                is_refraction = false;
-                break;
-            }
+            const GGXMDF mdf;
+            is_refraction = do_sample(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Move variable declaration here (and make const).

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 3]
```diff
@@ -2117,6 +2120,43 @@ namespace
                 }
             }
         }
+
+        static void update_assemblies(const AssemblyContainer& assemblies)
+        {
+            for (const auto& assembly: assemblies)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Insert space after `:`.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 4]
```diff
@@ -2117,6 +2120,43 @@ namespace
...
+            {
+                update_bsdfs(assembly.bsdfs());
+                update_assemblies(assembly.assemblies());
+            }
+        }
+
+        static void update_bsdfs(BSDFContainer& bsdfs)
+        {
+            for (auto& bsdf: bsdfs)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Here too.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 5]
```diff
@@ -2117,6 +2120,43 @@ namespace
...
+        {
+            ParamArray& params = bsdf.get_parameters();
+
+            const string mdf = params.get_optional<string>("mdf", "");
+            params.strings().remove("mdf");
+
+            if (mdf != "ggx")
+            {
+                RENDERER_LOG_WARNING(
+                    "The %s microfacet distribution used by BSDF %s was removed; "
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Until we revisit all log messages, replace the first `The` by `the`.

Comment (dictoon) [reply src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Quote entity names: `\"%s\"`

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 6]
```diff
@@ -893,26 +891,19 @@ namespace
...
                 CLOSURE_FLOAT_PARAM(Params, anisotropy),
                 CLOSURE_FLOAT_KEYPARAM(Params, energy_compensation, "energy_compensation"),
                 CLOSURE_FINISH_PARAM(Params)
             };
 
             shading_system.register_closure(name(), id(), params, &prepare_closure, nullptr);
-
             g_closure_convert_funs[id()] = &convert_closure;
-
             g_closure_get_modes_funs[id()] = &modes;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: `g_closure_get_modes_funs` still needed?

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 7]
```diff
@@ -2117,6 +2120,43 @@ namespace
...
+        }
+
+        static void update_microfacet_params(BSDF& bsdf)
+        {
+            ParamArray& params = bsdf.get_parameters();
+
+            const string mdf = params.get_optional<string>("mdf", "");
+            params.strings().remove("mdf");
+
+            if (mdf != "ggx")
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: If an `mdf` param was not found, `mdf` is empty here and we will (incorrectly) print the message. Better:

    if (params.strings().exist("mdf"))
    {
        const string mdf = params.strings().get<string>("mdf");
        params.strings().remove("mdf");
        if (mdf != "ggx")
            ...
    }

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 8]
```diff
@@ -679,13 +675,8 @@ namespace
...
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Same question here.

File: src/appleseed/renderer/kernel/shading/closures.cpp (hunk 1 above)
Comment (est77) [reply src/appleseed/renderer/kernel/shading/closures.cpp]: Yes, we use it to pick closures and when computing pdfs

**Pull request merged**